jax/_src/extend/ffi.py

```diff
@@ -15,11 +15,103 @@
 from __future__ import annotations
 
 import os
+from collections.abc import Iterable, Mapping, Sequence
+from typing import Any
 
+import numpy as np
+
+from jax._src import core
+from jax._src import dtypes
+from jax._src.interpreters import mlir
 from jax._src.lib import jaxlib
+from jax._src.lib.mlir import ir
+from jax._src.typing import DimSize
 
 
 def include_dir() -> str:
   """Get the path to the directory containing header files bundled with jaxlib"""
   jaxlib_dir = os.path.dirname(os.path.abspath(jaxlib.__file__))
   return os.path.join(jaxlib_dir, "include")
+
+
+def ffi_lowering(
+    call_target_name: str,
+    *,
+    operand_layouts: Sequence[Sequence[DimSize]] | None = None,
+    result_layouts: Sequence[Sequence[DimSize]] | None = None,
+    backend_config: Mapping[str, ir.Attribute] | None = None,
+    **lowering_args: Any
+) -> mlir.LoweringRule:
+  """Build a lowering rule for an foreign function interface (FFI) target
+
+  By default, this lowering rule can use the input and output abstract values to
+  compute the input and output types and shapes for the custom call, assuming
+  row-major layouts.
+
+  If keyword arguments are passed at bind time, these are treated as attributes,
+  and added to the `backend_config`.
+
+  Args:
+    call_target_name: The name of the custom call target.
+    operand_layouts: A sequence of layouts (dimension orders) for each operand.
+      By default, the operands are assumed to be row-major.
+    result_layouts: A sequence of layouts (dimension orders) for each result.
+      By default, the results are assumed to be row-major.
+    backend_config: Configuration data for the custom call. Any keyword
+      arguments provided at bind time will added to this dictionary.
+    lowering_args: Any other arguments to :func:`mlir.custom_call` will also be
+      passed through if provided as extra arguments to this function.
+  """
+
+  def _lowering(
+    ctx: mlir.LoweringRuleContext, *operands: ir.Value, **params: Any
+  ) -> Sequence[ir.Value | Sequence[ir.Value]]:
+    kwargs = dict(lowering_args)
+    kwargs["api_version"] = kwargs.get("api_version", 4)
+    kwargs["backend_config"] = dict(
+      backend_config or {}, **{k: _ir_attribute(v) for k, v in params.items()})
+
+    # TODO(dfm): This is a common pattern for supporting dynamic shapes, but I
```

Comment:
Not sure if this is really necessary here. CC @gnecula.

Reply:
I've just removed this. We can revisit in the future if it's a feature that many users would find useful.